src/middle/Middle.ml

```diff
@@ -131,6 +149,7 @@ let gensym_enter () =
   let old_counter = !_counter in
   (gensym (), fun () -> _counter := old_counter)
 
+let gensym_reset_danger_use_cautiously () = _counter := 0
```

Comment:
Without digging into the use cases, I'm wondering if this is necessary because there is also the gensym_enter function that returns a reset function that restores the old state of the counter.

Reply:
This is only meant for unit tests for optimizations. The idea is to reset it at the start of a unit test, so the results of the unit test are not dependent on the order that the unit tests are run in. Maybe it should have something like: only for use in testing in its name or doc?

I guess one could also use gensym_enter before running the test and then call reset again after the test is done. You're kind of relying on someone not forgetting to reset though and it seems harder to remember to write two things, rather than one, but don't feel so strongly about this.

Reply:
Hmm, gotcha. This makes sense, though ordinarily you'd use a test fixture for this. We might even consider making a monad for it (tagging @enetsee :P)